src/main/flight/pid.c

```diff
@@ -538,6 +567,79 @@ static void handleItermRotation()
     }
 }
 
+#ifdef USE_ACRO_TRAINER
+
+int8_t acroTrainerSign(float x)
+{
+    return (x > 0.0f) - (x < 0.0f);
+}
+
+// Acro Trainer - Manipulate the setPoint to limit axis angle while in acro mode
+// There are three states:
+// 1. Current angle has exceeded limit
+//    Apply correction to return to limit (similar to pidLevel)
+// 2. Future overflow has been projected based on current angle and gyro rate
+//    Manage the setPoint to control the gyro rate as the actual angle  approaches the limit (try to prevent overshoot)
+// 3. If no potential overflow is detected, then return the original setPoint
+
+static float applyAcroTrainer(int axis, const rollAndPitchTrims_t *angleTrim, float setPoint)
+{
+    float ret = setPoint;
+
+#ifndef SIMULATOR_BUILD
+    if (!FLIGHT_MODE(ANGLE_MODE) && !FLIGHT_MODE(HORIZON_MODE)) {
+        bool resetIterm = false;
+        float projectedAngle;
+        const int8_t setpointSign = acroTrainerSign(setPoint);
+        const float currentAngle = (attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f;
+        const int8_t angleSign = acroTrainerSign(currentAngle);
+
+        // Limit and correct the angle when it exceeds the limit
+        if ((fabsf(currentAngle) > acro_trainer_angle_limit) || (acroTrainerAxisState[axis] != 0)) {
+            if (angleSign == -setpointSign) {  // stick has reversed - stop limiting
+                acroTrainerAxisState[axis] = 0;
+            } else {
+                if (acroTrainerAxisState[axis] != angleSign) {
```

Comment:
Hmmm... Have to think about that.  Wasn't really expecting the oscillation to be so severe to make the angle cross zero in the reverse direction. Haven't seen that in any of my testing.

Actually I think that's a bug I introduced in rewriting it about 7 millions times! ;)  Should probably be:

`if (acroTrainerAxisState[axis] == -setpointSign)`

Since we care that the stick has moved to the opposite direction from the one that triggered the overflow in the first place.

Reply:
`angleSign == -setpointSign` should handle it (at least eventually if angle is opposite to AxisState)